crates/router/src/core/payments.rs

```diff
@@ -1211,55 +1198,35 @@ pub async fn list_payments(
                 )
                 .await
             {
-                Ok(pa) => pa,
-                Err(e) => return Err(e),
-            };
-            let mut map_write_locks = match internal_map.write() {
-                //Blocking wait for write
-                Ok(lock) => lock,
-                Err(_) => {
-                    logger::warn!("Failed to get lock in async context.");
-                    return Ok(()); // Exiting the async context
-                }
-            };
-            if map_write_locks.get(pi.payment_id.as_str()).is_none() {
-                map_write_locks.insert(&pi.payment_id, vec![pa]);
-            } else {
-                match map_write_locks.get_mut(pi.payment_id.as_str()) {
-                    Some(existing_pas) => {
-                        (*existing_pas).push(pa);
-                    }
-                    None => {
-                        return Ok(()); //Exiting async function
-                    }
+                Ok(pa) => return Some((pi.payment_id.as_str(), pa)),
+                Err(e) => {
+                    logger::warn!("Error getting payment_attempts using payment_id : {}", e);
+                    None
                 }
             }
-            Ok(())
-        })
-        .collect::<Vec<_>>();
-    let _result = join_all(collected_futures).await;
-    let mut pi_pa_tuple_vec: Vec<(storage::PaymentIntent, storage::PaymentAttempt)> = Vec::new();
-    let read_lock = match payment_intent_attempt_map.read() {
-        Ok(read_lock) => read_lock,
-        Err(_) => {
-            logger::warn!("Failed to gain read locks.");
-            return Err(errors::ApiErrorResponse::InternalServerError).into_report();
         }
-    };
+    });
 
-    for v in payment_intents.iter() {
-        let pa_list = match read_lock.get(v.payment_id.as_str()) {
-            Some(list) => list,
-            None => {
-                continue;
+    //Collecting PaymentAttempt of each PaymentIntent in hashmap
+    let mut pi_pa_map = join_all(collected_futures)
+        .await
+        .into_iter()
+        .flatten()
+        .collect::<HashMap<&str, storage::PaymentAttempt>>();
+
+    //Iterating over sorted Intents and sorting Attempts
+    let pi_pa_tuple_vec: Vec<(storage::PaymentIntent, storage::PaymentAttempt)> = payment_intents
+        .iter()
+        .filter_map(|v| {
+            //usinf for_each to get ownership of values
+            match pi_pa_map.remove(v.payment_id.as_str()) {
+                Some(pa) => Some((v.clone(), pa)),
```

Comment:
I still this clone should be avoided, that was the entire reason behind using iter and maps for this operation, what do you think? @lsampras

Reply:
Cool,
can convert this `.iter()` to `.into_iter()` as well
Don't see usages of payment_intents after this

Reply:
Also just realized that since we are calling collect on the futures iterator It automatically is converted to [FuturesUnordered](https://rust-lang-nursery.github.io/futures-api-docs/0.1.27/futures/stream/struct.FuturesUnordered.html)...

We might not need to worry about ordering since collecting the futures should lead to ordered results (as compared to calling poll on FuturesUnordered which can return a random result)
So in the initial case instead of returning `(payment_id, attempt_id)` & then sorting them out we can just return a vec of `(PaymentIntent, PaymentAttempt)` and call it a day.... (might need to verify this tho)